File: modules/mira-face-detector/android/src/main/java/expo/modules/mirafacedetector/MiraFaceDetectorPlugin.kt
package expo.modules.mirafacedetector

import android.media.Image
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetector
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class MiraFaceDetectorPlugin(proxy: VisionCameraProxy, options: Map<String, Any>?) : FrameProcessorPlugin() {

    // Throttle: ~5 fps (200ms between detections)
    private var lastProcessedTime: Long = 0
    private val minIntervalMs: Long = 200

    // ML Kit singleton face detector with FAST performance mode
    private val detector: FaceDetector

    init {
        val detectorOptions = FaceDetectorOptions.Builder()
            .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
            .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
            .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
            .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
            .setMinFaceSize(0.15f)
            .enableTracking()
            .build()

        detector = FaceDetection.getClient(detectorOptions)
    }

    override fun callback(frame: Frame, arguments: Map<String, Any>?): Any {
        val currentTime = System.currentTimeMillis()

        // Throttle: skip if called too soon
        if (currentTime - lastProcessedTime < minIntervalMs) {
            return createEmptyResult()
        }
        lastProcessedTime = currentTime

        // Get media image from frame safely
        val mediaImage: Image
        try {
            mediaImage = frame.image
        } catch (e: Exception) {
            return createEmptyResult()
        }

        // Get rotation from imageProxy
        val rotationDegrees: Int
        try {
            rotationDegrees = frame.imageProxy.imageInfo.rotationDegrees
        } catch (e: Exception) {
            return createEmptyResult()
        }

        val inputImage = InputImage.fromMediaImage(mediaImage, rotationDegrees)

        // Synchronous detection using latch with timeout
        val latch = CountDownLatch(1)
        var detectedFaces: List<Face>? = null
        var error: Exception? = null

        detector.process(inputImage)
            .addOnSuccessListener { faces ->
                detectedFaces = faces
                latch.countDown()
            }
            .addOnFailureListener { e ->
                error = e
                latch.countDown()
            }

        // Wait up to 150ms for result (safety timeout)
        val completed = latch.await(150, TimeUnit.MILLISECONDS)

        if (!completed || error != null || detectedFaces == null) {
            return createEmptyResult()
        }

        val faces = detectedFaces!!

        if (faces.isEmpty()) {
            return mapOf(
                "hasFace" to false,
                "facesCount" to 0
            )
        }

        // Get primary face (largest bounding box)
        val primaryFace = faces.maxByOrNull { it.boundingBox.width() * it.boundingBox.height() } ?: faces[0]
        val bounds = primaryFace.boundingBox

        val primaryData = mutableMapOf<String, Any>(
            "bounds" to mapOf(
                "x" to bounds.left,
                "y" to bounds.top,
                "w" to bounds.width(),
                "h" to bounds.height()
            ),
            "yaw" to primaryFace.headEulerAngleY,
            "pitch" to primaryFace.headEulerAngleX,
            "roll" to primaryFace.headEulerAngleZ
        )

        // Add classification probabilities if available
        primaryFace.smilingProbability?.let { prob ->
            primaryData["smilingProb"] = prob
        }
        primaryFace.leftEyeOpenProbability?.let { prob ->
            primaryData["leftEyeOpenProb"] = prob
        }
        primaryFace.rightEyeOpenProbability?.let { prob ->
            primaryData["rightEyeOpenProb"] = prob
        }

        return mapOf(
            "hasFace" to true,
            "facesCount" to faces.size,
            "primary" to primaryData,
            "frameWidth" to frame.width,
            "frameHeight" to frame.height
        )
    }

    private fun createEmptyResult(): Map<String, Any> {
        return mapOf(
            "hasFace" to false,
            "facesCount" to 0
        )
    }
}

File: modules/mira-face-detector/ios/MiraFaceDetectorPlugin.swift
import VisionCamera
import MLKitFaceDetection
import MLKitVision

@objc(MiraFaceDetectorPlugin)
public class MiraFaceDetectorPlugin: FrameProcessorPlugin {

    // Throttle: ~5 fps (200ms between detections)
    private var lastProcessedTime: CFTimeInterval = 0
    private let minIntervalSec: CFTimeInterval = 0.2

    // ML Kit face detector
    private let faceDetector: FaceDetector

    public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
        // Configure face detector for fast performance
        let detectorOptions = FaceDetectorOptions()
        detectorOptions.performanceMode = .fast
        detectorOptions.landmarkMode = .none
        detectorOptions.contourMode = .none
        detectorOptions.classificationMode = .all
        detectorOptions.minFaceSize = 0.15
        detectorOptions.isTrackingEnabled = true

        faceDetector = FaceDetector.faceDetector(options: detectorOptions)

        super.init(proxy: proxy, options: options)
    }

    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
        let currentTime = CACurrentMediaTime()

        // Throttle: skip if called too soon
        if currentTime - lastProcessedTime < minIntervalSec {
            return createEmptyResult()
        }
        lastProcessedTime = currentTime

        guard let pixelBuffer = CMSampleBufferGetImageBuffer(frame.buffer) else {
            return createEmptyResult()
        }

        // Create ML Kit vision image
        let visionImage = VisionImage(buffer: frame.buffer)
        visionImage.orientation = getImageOrientation(frame: frame)

        // Synchronous detection
        var detectedFaces: [Face]?
        let semaphore = DispatchSemaphore(value: 0)

        faceDetector.process(visionImage) { faces, error in
            if error == nil {
                detectedFaces = faces
            }
            semaphore.signal()
        }

        // Wait up to 100ms
        _ = semaphore.wait(timeout: .now() + 0.1)

        guard let faces = detectedFaces else {
            return createEmptyResult()
        }

        let facesCount = faces.count

        if facesCount == 0 {
            return [
                "hasFace": false,
                "facesCount": 0
            ] as [String: Any]
        }

        // Get primary face (largest)
        let primaryFace = faces.max(by: {
            $0.frame.width * $0.frame.height < $1.frame.width * $1.frame.height
        }) ?? faces[0]

        let bounds = primaryFace.frame

        var primaryData: [String: Any] = [
            "bounds": [
                "x": bounds.origin.x,
                "y": bounds.origin.y,
                "w": bounds.width,
                "h": bounds.height
            ] as [String: Any]
        ]

        // Add angles
        primaryData["yaw"] = primaryFace.headEulerAngleY
        primaryData["pitch"] = primaryFace.headEulerAngleX
        primaryData["roll"] = primaryFace.headEulerAngleZ

        // Add classification probabilities if available
        if primaryFace.hasLeftEyeOpenProbability {
            primaryData["leftEyeOpenProb"] = primaryFace.leftEyeOpenProbability
        }
        if primaryFace.hasRightEyeOpenProbability {
            primaryData["rightEyeOpenProb"] = primaryFace.rightEyeOpenProbability
        }
        if primaryFace.hasSmilingProbability {
            primaryData["smilingProb"] = primaryFace.smilingProbability
        }

        // Get frame dimensions
        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)

        return [
            "hasFace": true,
            "facesCount": facesCount,
            "primary": primaryData,
            "frameWidth": width,
            "frameHeight": height
        ] as [String: Any]
    }

    private func createEmptyResult() -> [String: Any] {
        return [
            "hasFace": false,
            "facesCount": 0
        ]
    }

    private func getImageOrientation(frame: Frame) -> UIImage.Orientation {
        switch frame.orientation {
        case .portrait:
            return .right
        case .portraitUpsideDown:
            return .left
        case .landscapeLeft:
            return .up
        case .landscapeRight:
            return .down
        @unknown default:
            return .up
        }
    }
}
